# DP: argparse.py: Make the gettext import conditional

Index: python3.10-3.10.0~b1/Lib/argparse.py
===================================================================
--- python3.10-3.10.0~b1.orig/Lib/argparse.py
+++ python3.10-3.10.0~b1/Lib/argparse.py
@@ -89,7 +89,16 @@ import os as _os
 import re as _re
 import sys as _sys
 
-from gettext import gettext as _, ngettext
+try:
+    from gettext import gettext as _, ngettext
+except ImportError:
+    def _(message):
+        return message
+    def ngettext(singular,plural,n):
+        if n == 1:
+            return singular
+        else:
+            return plural
 
 SUPPRESS = '==SUPPRESS=='
 
